### ISSUE TITLE
Terraform failing EKS resources delete due to a wrong deletion order / missing retry in Terraform Test

### ISSUE THREAD

**Issue by cristianocasella** [hashicorp/terraform#29732]
Hello everyone,
I'm trying to use Terraform Test with my EKS cluster management module.

The module is taking care of the following object on the network side:

- VPC
- Subnets
- Internet Gateway
- Routing tables

And on the Kubernetes side:

 - Kubernetes Cluster
 - Kubernetes Node Group
 - IAM Roles
 
### Current Behaviour
The creation of the resources and the test execution is performed correctly, during the deletion of the resource there is a failure (sanitized output):
 
> Warning: The "terraform test" command is experimental
>
> We'd like to invite adventurous module authors to write integration tests
> for their modules using this command, but all of the behaviors of this
> command are currently experimental and may change based on feedback.
> 
> For more information on the testing experiment, including ongoing research
> goals and avenues for feedback, see:
>     https://www.terraform.io/docs/language/modules/testing-experiment.html
>
>
> Error: Failed to clean up after tests
> 
> Due to errors during destroy, test suite "defaults" has left behind an
> object for
> module.main.kubernetes_secret.container_registry_secret. You
> will need to delete this object manually in the remote system, or else it
> may have an ongoing cost.
>
>
> Error: Failed to clean up after tests
> 
> Due to errors during destroy, test suite "defaults" has left behind an
> object for module.main.kubernetes_secret.credentials. You will need
> to delete this object manually in the remote system, or else it may have an
> ongoing cost.
>
> Error: Failed to clean up after tests
>
> Due to errors during destroy, test suite "defaults" has left behind an
> object for module.main.aws_eks_cluster.k8s_cluster, with the following
> identity:
>    name = "main-cluster"
>
> You will need to delete this object manually in the remote system, or else
> it may have an ongoing cost.
>
> Error: Failed to clean up after tests
> 
> Due to errors during destroy, test suite "defaults" has left behind an
> object for module.main.aws_eks_node_group.k8s_cluster_node_group. You will
> need to delete this object manually in the remote system, or else it may
> have an ongoing cost.
>
> Error: Failed to clean up after tests
>
> Due to errors during destroy, test suite "defaults" has left behind an
> object for
> module.main.kubernetes_namespace.project-namespace. You
> will need to delete this object manually in the remote system, or else it
> may have an ongoing cost.
>
> ─── Error: defaults (terraform destroy) ─────────────────────────────────────
> failed to destroy objects created during test (NOTE: leftover remote objects
> may still exist)
>
> Warning: The test provider is experimental
> 
>   with provider["terraform.io/builtin/test"],
> The Terraform team is using the test provider (terraform.io/builtin/test)
> as part of ongoing research about declarative testing of Terraform modules.
> 
> The availability and behavior of this provider is expected to change
> significantly even in patch releases, so we recommend using this provider
> only in test configurations and constraining your test configurations to an
> exact Terraform version.
>
> Warning: The test provider is experimental
> 
>  with provider["terraform.io/builtin/test"],
> The Terraform team is using the test provider (terraform.io/builtin/test)
> as part of ongoing research about declarative testing of Terraform modules.
>
> The availability and behavior of this provider is expected to change
> significantly even in patch releases, so we recommend using this provider
> only in test configurations and constraining your test configurations to an
> exact Terraform version.
>
> Error: Missing pending object in plan
>
>   on main.tf line 42:
>   (source code not available)
>
> Instance module.main.aws_eks_cluster.k8s_cluster is marked as having a
> change pending but that change is not recorded in the plan. This is a bug
> in Terraform; please report it.
>
> Error: Missing pending object in plan
>
>  on main.tf line 42:
>   (source code not available)
>
> Instance module.main.aws_eks_cluster.k8s_cluster is marked as having a
> change pending but that change is not recorded in the plan. This is a bug
> in Terraform; please report it.

The deletion process is trying to delete the k8s cluster, but the nodegroup is still present, so it fails. After the nodegroup deletion, there is no more attempt to delete the k8s cluster, that it's preventing now the deletion of the network objects where the cluster is currently installed.
This is happening also with the nodegroup definition having the k8s cluster as dependency.

This failure is happening only during the Terraform Test. In a normal usage of the module, the resources can be properly created and destroyed.

### Expected behaviour
After the nodegroup deletion a new attempt of deleting the K8s cluster should happens.

**Comment by apparentlymart** [comment]
I did notice notice a couple interesting things from a little more skimming before I finish for the day:

* This new trace log of `terraform test` doesn't seem to include the "Error: Missing pending object in plan" error we saw in the original issue. Instead, it's returning more reasonable errors that are indicative of failing real API calls:

    ```
    ╷
    │ Error: error deleting EKS Cluster (testing-cluster): ResourceInUseException: Cluster has nodegroups attached
    │ {
    │   RespMetadata: {
    │     StatusCode: 409,
    │     RequestID: "5e921bb0-4bd3-4ac8-b560-eeb82ad82114"
    │   },
    │   ClusterName: "testing-cluster",
    │   Message_: "Cluster has nodegroups attached",
    │   NodegroupName: "testing-cluster_node_group"
    │ }
    │ 
    │ 
    ╵
    ╷
    │ Error: error deleting subnet (subnet-045473a6dded7cf6b): timeout while waiting for state to become 'destroyed' (last state: 'pending', timeout: 20m0s)
    │ 
    │ 
    ╵
    ╷
    │ Error: error deleting subnet (subnet-057ea6dc8dc171d48): timeout while waiting for state to become 'destroyed' (last state: 'pending', timeout: 20m0s)
    │ 
    │ 
    ╵
    ╷
    │ Error: Error deleting VPC: DependencyViolation: The vpc 'vpc-08a67f6b5bfedaf58' has dependencies and cannot be deleted.
    │ 	status code: 400, request id: 9faee4f7-8280-4cc1-a715-492b37b08c15
    │ 
    │ 
    ╵
    ```
* I guess this new output is from "the network side" whereas the original report was from "the kubernetes side", which are two separate modules under test. Is that right, @cristianocasella? No problem if so, I just want to make sure we can understand how these new results relate to what you included originally.

This new behavior does indeed seem like a regular old "incorrect ordering" problem, rather than a more fundamental bug as I suspected with the "missing pending object" errors. There might be something else lurking under here that was causing those other errors, but I think for now we can focus our attention on whether the dependency graph here was correct and, if so, whether Terraform actually respected it. It could be helpful to also see the `TF_LOG=trace` result for the `terraform -chdir=test/defaults destroy` run which worked, since then we could compare the final apply graph with what was generated for `terraform test` and see if there are any differences, which could therefore help give us some leads to narrow down where the problem might be.

**Comment by jbardin** [comment]
After coming back here to do some more debugging, I'm finding that this is still working correctly, but I verified the intermediate state is correct, the graph is correct, and during destroy the following message is shown:

```
"module.main.aws_eks_cluster.k8s_cluster (destroy)" is waiting for "module.main.aws_eks_node_group.k8s_cluster_node_group (destroy)"
```

This is with the current main branch, so I have a feeling that some state handling changes we made have fixed a bug in the test harness. We have a v1.1-beta being published shortly, and I suggest giving that a try to verify your configuration as well.

Thanks!
